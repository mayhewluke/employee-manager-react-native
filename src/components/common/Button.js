import React from "react";
import { TouchableOpacity, Text } from "react-native";
import PropTypes from "prop-types";

const styles = {
  buttonStyle: {
    flex: 1,
    alignSelf: "stretch",
    backgroundColor: "#fff",
    borderRadius: 5,
    borderWidth: 1,
    borderColor: "#007aff",
    marginLeft: 5,
    marginRight: 5
  },
  textStyle: {
    alignSelf: "center",
    color: "#007aff",
    fontSize: 16,
    fontWeight: "600",
    paddingTop: 10,
    paddingBottom: 10
  }
};

const Button = ({ onPress, children }) => {
  const { buttonStyle, textStyle } = styles;
  return (
    <TouchableOpacity style={buttonStyle} onPress={onPress}>
      <Text style={textStyle}>{children}</Text>
    </TouchableOpacity>
  );
};

Button.propTypes = {
  onPress: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired
};

export { Button };
