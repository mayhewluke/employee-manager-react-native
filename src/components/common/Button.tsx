import React from "react";
import { StyleSheet, Text, TouchableOpacity } from "react-native";

const styles = StyleSheet.create({
  buttonStyle: {
    backgroundColor: "#fff",
    borderColor: "#007aff",
    borderRadius: 5,
    borderWidth: 1,
    marginLeft: 5,
    marginRight: 5,
  },
  textStyle: {
    alignSelf: "center",
    color: "#007aff",
    fontSize: 16,
    fontWeight: "600",
    paddingBottom: 10,
    paddingTop: 10,
  },
});

const Button: React.SFC<{ onPress: () => void }> = ({ onPress, children }) => {
  const { buttonStyle, textStyle } = styles;
  return (
    <TouchableOpacity style={buttonStyle} onPress={onPress}>
      <Text style={textStyle}>{children}</Text>
    </TouchableOpacity>
  );
};

export default Button;
