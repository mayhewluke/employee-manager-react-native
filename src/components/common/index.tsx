export { default as AsyncButton } from "./AsyncButton";
export { default as Button } from "./Button";
export { default as Card } from "./Card";
export { default as CardSection } from "./CardSection";
export { default as Confirm } from "./Confirm";
export { default as Input } from "./Input";
export { default as Spinner, Size as SpinnerSize } from "./Spinner";
