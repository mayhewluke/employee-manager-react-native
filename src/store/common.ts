export type Async<T> =
  | { state: "INIT" }
  | { state: "PROGRESS" }
  | { state: "COMPLETE"; value: T }
  | { state: "ERROR"; error: string };
