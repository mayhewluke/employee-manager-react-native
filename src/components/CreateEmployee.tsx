import React, { Component } from "react";
import { Picker, StyleSheet, Text } from "react-native";
import { NavigationScreenProp, NavigationScreenProps } from "react-navigation";
import { connect } from "react-redux";

import { IRootState } from "../store";
import {
  createEmployee,
  FieldUpdatePayload,
  IEmployee,
  resetForm,
  ShiftDay,
  updateField,
} from "../store/Employee";
import {
  Button,
  Card,
  CardSection,
  Input,
  Spinner,
  SpinnerSize,
} from "./common";

const renderButton = (isLoading: boolean, onSubmit: () => void) => {
  if (isLoading) {
    return <Spinner size={SpinnerSize.Large} />;
  }
  return <Button onPress={onSubmit}>Create</Button>;
};

const styles = StyleSheet.create({
  errorTextStyle: {
    alignSelf: "center",
    color: "red",
    fontSize: 20,
  },
  pickerContainerStyle: {
    flexDirection: "column",
    paddingLeft: 25,
  },
  pickerLabelStyle: {
    fontSize: 18,
  },
});

interface IProps extends NavigationScreenProps {
  employeeName: string;
  error: string;
  loading: boolean;
  phone: string;
  shift: ShiftDay;
  dispatchFieldUpdate: (payload: FieldUpdatePayload) => void;
  dispatchCreateEmployee: (
    employee: IEmployee,
    navigation: NavigationScreenProp<any>,
  ) => void;
  dispatchResetForm: () => void;
}

class CreateEmployee extends Component<IProps> {
  public static navigationOptions = {
    title: "Add an employee",
  };

  public componentWillUnmount() {
    const { dispatchResetForm } = this.props;
    dispatchResetForm();
  }

  public render() {
    const {
      employeeName,
      phone,
      shift,
      dispatchFieldUpdate,
      dispatchCreateEmployee,
      error,
      loading,
      navigation,
    } = this.props;
    const dispatchUpdateName = (value: string) =>
      dispatchFieldUpdate({ value, field: "employeeName" });
    const dispatchUpdatePhone = (value: string) =>
      dispatchFieldUpdate({ value, field: "phone" });
    const dispatchUpdateShift = (value: ShiftDay) =>
      dispatchFieldUpdate({ value, field: "shift" });
    return (
      <Card>
        <CardSection>
          <Input
            label="Name"
            placeholder="Taylor"
            value={employeeName}
            onChangeText={dispatchUpdateName}
          />
        </CardSection>
        <CardSection>
          <Input
            label="Phone"
            placeholder="555-555-5555"
            value={phone}
            onChangeText={dispatchUpdatePhone}
          />
        </CardSection>

        <CardSection style={styles.pickerContainerStyle}>
          <Text style={styles.pickerLabelStyle}>Shift day</Text>
          <Picker selectedValue={shift} onValueChange={dispatchUpdateShift}>
            <Picker.Item label={ShiftDay.Monday} value={ShiftDay.Monday} />
            <Picker.Item label={ShiftDay.Tuesday} value={ShiftDay.Tuesday} />
            <Picker.Item
              label={ShiftDay.Wednesday}
              value={ShiftDay.Wednesday}
            />
            <Picker.Item label={ShiftDay.Thursday} value={ShiftDay.Thursday} />
            <Picker.Item label={ShiftDay.Friday} value={ShiftDay.Friday} />
            <Picker.Item label={ShiftDay.Saturday} value={ShiftDay.Saturday} />
            <Picker.Item label={ShiftDay.Sunday} value={ShiftDay.Sunday} />
          </Picker>
        </CardSection>

        {error !== "" && (
          <CardSection>
            <Text style={styles.errorTextStyle}>{error}</Text>
          </CardSection>
        )}

        <CardSection>
          {renderButton(loading, () =>
            dispatchCreateEmployee(
              {
                employeeName,
                phone,
                shift,
                uid: null,
              },
              navigation,
            ),
          )}
        </CardSection>
      </Card>
    );
  }
}

const mapStateToProps = (state: IRootState) => {
  const { employeeName, phone, shift, loading, error } = state.employee;
  return {
    employeeName,
    error,
    loading,
    phone,
    shift,
  };
};

export default connect(
  mapStateToProps,
  {
    dispatchCreateEmployee: createEmployee,
    dispatchFieldUpdate: updateField,
    dispatchResetForm: resetForm,
  },
)(CreateEmployee);
