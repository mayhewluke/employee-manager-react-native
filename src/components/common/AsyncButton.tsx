import * as React from "react";
import { StyleSheet, Text, View } from "react-native";

import { Async } from "../../store";
import Button from "./Button";
import Spinner, { Size as SpinnerSize } from "./Spinner";

interface IProps {
  label: string;
  asyncAction: Async<any>;
  onPress: () => void;
  size?: SpinnerSize;
}

const styles = StyleSheet.create({
  errorTextStyle: {
    alignSelf: "center",
    color: "red",
    fontSize: 20,
  },
});

const AsyncButton: React.SFC<IProps> = ({
  label,
  asyncAction,
  onPress,
  size = SpinnerSize.Large,
}) => (
  <View style={{ flex: 1, flexDirection: "row" }}>
    {asyncAction.state === "ERROR" && (
      <Text style={styles.errorTextStyle}>{asyncAction.error}</Text>
    )}
    {asyncAction.state === "PROGRESS" ? (
      <Spinner size={size} />
    ) : (
      <Button onPress={onPress}>{label}</Button>
    )}
  </View>
);

export default AsyncButton;
