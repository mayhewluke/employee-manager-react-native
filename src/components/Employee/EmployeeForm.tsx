import * as React from "react";
import { Picker, StyleSheet, Text, View } from "react-native";
import { connect } from "react-redux";

import { IRootState } from "../../store";
import {
  FieldUpdatePayload,
  ShiftDay,
  updateField,
} from "../../store/Employee";
import { CardSection, Input } from "../common";

const styles = StyleSheet.create({
  errorTextStyle: {
    alignSelf: "center",
    color: "red",
    fontSize: 20,
  },
  pickerContainerStyle: {
    flexDirection: "column",
    paddingLeft: 25,
  },
  pickerLabelStyle: {
    fontSize: 18,
  },
});

export interface IFormProps {
  employeeName: string;
  phone: string;
  shift: ShiftDay;
  dispatchFieldUpdate: (payload: FieldUpdatePayload) => void;
  error: string;
}

const Form = ({
  employeeName,
  phone,
  shift,
  dispatchFieldUpdate,
  error,
}: IFormProps) => {
  const dispatchUpdateName = (value: string) =>
    dispatchFieldUpdate({ value, field: "employeeName" });
  const dispatchUpdatePhone = (value: string) =>
    dispatchFieldUpdate({ value, field: "phone" });
  const dispatchUpdateShift = (value: ShiftDay) =>
    dispatchFieldUpdate({ value, field: "shift" });
  return (
    <View>
      <CardSection>
        <Input
          label="Name"
          placeholder="Taylor"
          value={employeeName}
          onChangeText={dispatchUpdateName}
        />
      </CardSection>
      <CardSection>
        <Input
          label="Phone"
          placeholder="555-555-5555"
          value={phone}
          onChangeText={dispatchUpdatePhone}
        />
      </CardSection>

      <CardSection style={styles.pickerContainerStyle}>
        <Text style={styles.pickerLabelStyle}>Shift day</Text>
        <Picker selectedValue={shift} onValueChange={dispatchUpdateShift}>
          {Object.keys(ShiftDay).map(key => (
            <Picker.Item
              key={key}
              value={key}
              label={ShiftDay[key as keyof typeof ShiftDay]}
            />
          ))}
        </Picker>
      </CardSection>

      {error !== "" && (
        <CardSection>
          <Text style={styles.errorTextStyle}>{error}</Text>
        </CardSection>
      )}
    </View>
  );
};

const mapStateToProps = (state: IRootState) => {
  const { employeeName, phone, shift } = state.employee;
  return { employeeName, phone, shift };
};

export default connect(
  mapStateToProps,
  { dispatchFieldUpdate: updateField },
)(Form);
