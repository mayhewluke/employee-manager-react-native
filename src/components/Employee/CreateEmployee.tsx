import React, { Component } from "react";
import { NavigationScreenProp, NavigationScreenProps } from "react-navigation";
import { connect } from "react-redux";

import { Async, IRootState } from "../../store";
import {
  createEmployee,
  IEmployee,
  resetForm,
  ShiftDay,
  updateField,
} from "../../store/Employee";
import { Button, Card, CardSection, Spinner, SpinnerSize } from "../common";
import EmployeeForm, { IFormProps } from "./EmployeeForm";

// TODO extract into a reusable component
const renderButton = (isLoading: boolean, onSubmit: () => void) => {
  if (isLoading) {
    return <Spinner size={SpinnerSize.Large} />;
  }
  return <Button onPress={onSubmit}>Create</Button>;
};

interface IProps extends IFormProps, NavigationScreenProps {
  employeeName: string;
  phone: string;
  shift: ShiftDay;
  createAction: Async<null>;
  dispatchCreateEmployee: (
    employee: IEmployee<null>,
    navigation: NavigationScreenProp<any>,
  ) => void;
  dispatchResetForm: () => void;
}

class CreateEmployee extends Component<IProps> {
  public static navigationOptions = {
    title: "Add an employee",
  };

  public componentWillUnmount() {
    const { dispatchResetForm } = this.props;
    dispatchResetForm();
  }

  public render() {
    const {
      employeeName,
      phone,
      shift,
      dispatchCreateEmployee,
      createAction,
      navigation,
    } = this.props;
    return (
      <Card>
        <EmployeeForm {...this.props} asyncAction={createAction} />

        <CardSection>
          {renderButton(createAction.state === "PROGRESS", () =>
            dispatchCreateEmployee(
              {
                employeeName,
                phone,
                shift,
                uid: null,
              },
              navigation,
            ),
          )}
        </CardSection>
      </Card>
    );
  }
}

const mapStateToProps = (state: IRootState) => {
  const { employeeName, phone, shift, createAction } = state.employee;
  return {
    createAction,
    employeeName,
    phone,
    shift,
  };
};

export default connect(
  mapStateToProps,
  {
    dispatchCreateEmployee: createEmployee,
    dispatchFieldUpdate: updateField,
    dispatchResetForm: resetForm,
  },
)(CreateEmployee);
