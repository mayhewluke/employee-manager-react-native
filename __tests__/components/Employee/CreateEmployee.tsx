import "jest-enzyme";
import React from "react";

import { shallow } from "enzyme";

import { NavigationScreenProps } from "react-navigation";

import { stubNavigation } from "../../helpers/react-navigation";
import { StubbedStore } from "../../helpers/redux";

import { AsyncButton } from "../../../src/components/common";
import CreateEmployee from "../../../src/components/Employee/CreateEmployee";
import EmployeeForm from "../../../src/components/Employee/EmployeeForm";
import {
  createEmployee,
  IEmployeeState,
  resetForm,
  ShiftDay,
  updateField,
} from "../../../src/store/Employee";

jest.mock("../../../src/store/Employee");
jest.mock(
  "../../../src/components/Employee/EmployeeForm",
  () => "EmployeeForm",
);

// For some reason TypeScript doesn't know about a `connect`ed class' static
// methods, and no solution is readily available after searching, so just
// override the types here until a better solution can be found.
const { navigationOptions } = CreateEmployee as any;

describe("CreateEmployee", () => {
  let props: NavigationScreenProps;
  let store: StubbedStore;
  let state: { employee: IEmployeeState };
  beforeEach(() => {
    state = {
      employee: {
        createAction: { state: "INIT" },
        employeeName: "Taylor",
        fireAction: { state: "INIT" },
        fireModalShown: false,
        phone: "555-5555",
        shift: ShiftDay.Monday,
        uid: null,
        updateAction: { state: "INIT" },
      },
    };
    store = new StubbedStore(state);
    props = {
      navigation: stubNavigation(),
    };
  });
  // This has side effects! Don't put this in beforeEach in case a test needs to
  // do something *before* running those side effects!
  const mkWrapper = () =>
    shallow(<CreateEmployee {...props} />, {
      context: { store },
    }).dive();

  it("renders properly", () => {
    expect(mkWrapper()).toMatchSnapshot();
  });

  it("passes dispatchFieldUpdate to the EmployeeForm", () => {
    const employeeForm = mkWrapper().find(EmployeeForm);
    // Can't just check the functions are equal since the dispatch* version is
    // wrapped inside another function in order to hook up the dispatch
    expect(updateField).toHaveBeenCalledTimes(0);

    employeeForm.props().dispatchFieldUpdate();
    expect(updateField).toHaveBeenCalledTimes(1);
  });

  it("uses an AsyncButton connected to createAction", () => {
    const button = mkWrapper().find({ label: "Create" });

    expect(button.type()).toBe(AsyncButton);
    expect(button.prop("asyncAction")).toBe(state.employee.createAction);
  });

  it("creates the employee when submitted", () => {
    const { employeeName, phone, shift } = state.employee;
    const button = mkWrapper().find({ label: "Create" });

    expect(createEmployee).not.toHaveBeenCalled();
    button.simulate("press");

    expect(createEmployee).toHaveBeenCalledTimes(1);
    expect(createEmployee).toHaveBeenCalledWith(
      { employeeName, phone, shift, uid: null },
      props.navigation,
    );
  });

  it("sets the page title", () => {
    expect(navigationOptions.title).toEqual("Add an employee");
  });

  it("resets the form when unmounted", () => {
    const wrapper = mkWrapper();
    expect(resetForm).not.toHaveBeenCalled();

    wrapper.unmount();

    expect(resetForm).toHaveBeenCalledTimes(1);
  });
});
