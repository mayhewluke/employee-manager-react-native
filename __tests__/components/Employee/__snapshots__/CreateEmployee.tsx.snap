// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEmployee renders properly 1`] = `
<Card>
  <EmployeeForm
    createAction={
      Object {
        "state": "INIT",
      }
    }
    dispatchCreateEmployee={[Function]}
    dispatchFieldUpdate={[Function]}
    dispatchResetForm={[Function]}
    employeeName="Taylor"
    navigation={
      Object {
        "addListener": [MockFunction],
        "closeDrawer": [MockFunction],
        "dismiss": [MockFunction],
        "dispatch": [MockFunction],
        "getParam": [MockFunction],
        "goBack": [MockFunction],
        "isFocused": [MockFunction],
        "navigate": [MockFunction],
        "openDrawer": [MockFunction],
        "pop": [MockFunction],
        "popToTop": [MockFunction],
        "push": [MockFunction],
        "replace": [MockFunction],
        "setParams": [MockFunction],
        "state": Object {},
        "toggleDrawer": [MockFunction],
      }
    }
    phone="555-5555"
    shift="Monday"
  />
  <CardSection>
    <AsyncButton
      asyncAction={
        Object {
          "state": "INIT",
        }
      }
      label="Create"
      onPress={[Function]}
    />
  </CardSection>
</Card>
`;
